### DISCUSSION TITLE
Render a markdown with typst using a template with arguments

### DISCUSSION THREAD

**Discussion by poc-chuc** [quarto-dev/quarto-cli#10425] [Q&A]
### Description

Hi,
I have a markdown file that I want to render using typst.
How can I set the front matter so it uses a typst template with arguments?

I am using Quarto in VS code, if it matters.

** edited typos since I wrote this from my tablet and now I have a proper keyword

**Comment by cderv** [comment]
Hi! 

Glad you are trying to work with Quarto ! 

What did you try so far ? Did you read our docs on how quarto is working and how templates are working ? 
Do you have an example of what you tried and what is not working as it should ? Or maybe explain more of what you tried and may not have understood ? 

Sharing a reproducible example is usually good manner if you want anyone passing by here to help with your project. Without a starter project to work with, I doubt you'll find anyone with the time to build one for you. This is my best advice based on my experience with how dev communique Q&A works best.

**Reply by poc-chuc** [reply]
Thanks for replying. 
I have gone over the Quarto and Typst documentation but I can't find the answer.

So, I have a report written in markdown, which I want to render to PDF. The report needs to have a cover page, headers, and footers.

I decided to start with the bubble typist template and then modify it with my own logo, etc.

From the documentation I know that I can state the typst template in the frontmatter, 
```
  formats:
    typst:
      template: template.typ
```

but how to do it when the template has arguments?

A typst file that users the template would look like this:

```
#import "@preview/bubble:0.2.0": *

#show: bubble.with(
  title: "Bubble template",
  subtitle: "Simple and colorful template",
  author: "hzkonor",
  affiliation: "University",
  date: datetime.today().display(),
  year: "Year",
  class: "Class",
  logo: image("logo.png"),
  color-words: ("important",)
) 
```
However, how can I call the `#show ` function  and provide the arguments from a markdown file?

**Reply by mcanouil** [reply]
Did you look at the "partials" code shared by Gordon?

- for example: https://github.com/quarto-dev/quarto-cli/blob/main/src/resources/formats/typst/pandoc/quarto/typst-show.typ

You need to do the same and use Pandoc templating system.

If you don't mean YAML options/parameters by "arguments" then you need to clarify what you are calling "arguments".

**Reply by poc-chuc** [reply]
I am a bit lost. Is you recommendation to download the typst pandoc template and start modifying from there?

**Reply by mcanouil** [reply]
That's the partials.
See <https://quarto.org/docs/journals/templates.html#typst-partials>.
Look at the description of the files.

If you don't use show, how do you expect your template to use metadata that are specific to your template?

Everything you need is described in the links shared by Gordon.
I suggest you to have a second reading.

**Reply by poc-chuc** [reply]
That is what I am trying to find out by creating this post

**Reply by mcanouil** [reply]
You need to use/define the Typst-show.typ partial as documented in the custom Typst format page of the documentation.

From the partials documentation page:

> typst-show.typ
> A show rule for the entire document that captures document metadata and passes it to the function defined in typst-template.typ.

From the custom Typst format page
![image](https://github.com/user-attachments/assets/f97d81b5-f441-4aac-91ec-0d084c3710c5)

Sorry but what is not clear with the need to use this partial (or its logic) in addition to the "template" to get more metadata into it?

**Reply by mcanouil** [reply]
If you use "template" and as documented it means you want to override the whole template.
It means you have to implement the logic from the partials yourself in the template.

- https://github.com/quarto-dev/quarto-cli/blob/main/src/resources/formats/typst/pandoc/quarto/template.typ

See the template syntax documentation if looking at the files is not enough: https://pandoc.org/MANUAL.html#template-syntax

**Reply by poc-chuc** [reply]
Thanks. I think I have it now:
If `typst-template.typ` defines:

    #let bubble(
        title: "",
        subtitle: "",
        author: "",
        affiliation: "",
    and so on

then `typst-show.typ` should look like:

    #show: doc => bubble(
        $if(title)$
            title: [$title$],
        $endif$
        $if(subtitle)$
          subtitle: [$subtitle$],
        $endif$
        and so on

and finally the frontmatter should look like:

    title: "Section 141096"
    subtitle: "My subtitle"
    formats:
        typst:
            template-partials:
                - typst-template.typ
                - typst-show.typ

Thanks again

**Comment by gordonwoodhull** [comment]
If I understand the question correctly, you have a Typst template that you want to use with Quarto. (It's good to specify "Typst template" because there are also Pandoc templates, which can lead to confusion.)

Currently you need to turn this into Quarto Typst partials:
https://quarto.org/docs/journals/templates.html#typst-partials

You can also check the existing Typst custom formats to see if your Typst template, or a similar one,  has been turned into partials already:

https://quarto.org/docs/output-formats/typst-custom.html

(I'd like a way to skip this step and use a Typst template directly, but I'm not sure if this is possible.)

**Reply by poc-chuc** [reply]
Thank you. You are correct, I was referring to a typst template. I went into more detail in my answer above.
